### PR TITLE
Optional packet filter

### PR DESCRIPTION
**Pull request by k-danil**
Good day!

This PR introduce filtering capabilities on packet parsing step:
- `DemuxerOptPacketFilter` is used to pass user defined closure `type PacketFilter func(p Packet) (skip bool)`
- End user can access all of PacketHeader and PacketAdaptationField fields to make a decision
- NextPacket() will iterate packets until we get packet or error, so filter do not brake current behaviour

### CODE REVIEW

**Comment by asticode** [comment]
Thanks a lot for all the PRs!! ❤️ 

I'll take a look as soon as I have free time 👍

**Comment by k-danil** [comment]
Thank you so much!

File: demuxer.go [hunk 1]
```diff
@@ -23,21 +23,28 @@ var (
...
+	r            io.Reader
 }
 
 // PacketsParser represents an object capable of parsing a set of packets containing a unique payload spanning over those packets
 // Use the skip returned argument to indicate whether the default process should still be executed on the set of packets
 type PacketsParser func(ps []*Packet) (ds []*DemuxerData, skip bool, err error)
 
+// PacketFilter executed right before pushing byte-payload to Packet-object. PacketHeader and PacketAdaptationField still be present
+// Use the skip to filter-out unwanted packets from pipeline. NextPacket() will return next unfiltered packet in query
+type PacketFilter func(p *Packet) (skip bool)
```
Comment (asticode) [inline demuxer.go:46]: Could you rename this to `PacketSkipper`?

Also could you change the above comment to:
```PacketSkipper represents an object capable of skipping a packet before parsing its payload. Its header and adaptation field is parsed and provided to the object.
Use this option if you need to filter out unwanted packets from your pipeline. NextPacket() will return the next unfiltered packet if any.```

File: demuxer.go [hunk 2]
```diff
@@ -23,21 +23,28 @@ var (
...
+	dataBuffer []*DemuxerData
+	l          astikit.CompleteLogger
+
 	optPacketSize    int
 	optPacketsParser PacketsParser
-	packetBuffer     *packetBuffer
-	packetPool       *packetPool
-	programMap       *programMap
-	r                io.Reader
+	optPacketFilter  PacketFilter
```
Comment (asticode) [inline demuxer.go:32]: Could you rename the variable to `optPacketSkipper`?

File: demuxer.go [hunk 3]
```diff
@@ -78,6 +85,13 @@ func DemuxerOptPacketsParser(p PacketsParser) func(*Demuxer) {
 	}
 }
 
+// DemuxerOptPacketFilter returns the option to set the packet filter
+func DemuxerOptPacketFilter(p PacketFilter) func(*Demuxer) {
```
Comment (asticode) [inline demuxer.go:89]: Could your rename the variable from `p` to `s`?

File: packet.go [hunk 4]
```diff
@@ -74,7 +74,7 @@ type PacketAdaptationExtensionField struct {
 }
 
 // parsePacket parses a packet
-func parsePacket(i *astikit.BytesIterator) (p *Packet, err error) {
+func parsePacket(i *astikit.BytesIterator, flt PacketFilter) (p *Packet, err error) {
```
Comment (asticode) [inline packet.go:77]: Could your rename the variable from `flt` to `s` ?

File: packet.go [hunk 5]
```diff
@@ -109,6 +109,11 @@ func parsePacket(i *astikit.BytesIterator) (p *Packet, err error) {
 		}
 	}
 
+	// Use optional packet filter
+	if flt != nil && flt(p) {
+		return nil, nil
```
Comment (asticode) [inline packet.go:114]: Could you add a `var errSkippedPacket = errors.New("astits: skipped packet")` at the top of this file and return `nil, errSkippedPacket` when packet is skipped?

I'd rather indicate to the caller that packet has been skipped.

File: packet_buffer.go [hunk 6]
```diff
@@ -11,15 +11,17 @@ import (
 // packetBuffer represents a packet buffer
 type packetBuffer struct {
 	packetSize       int
+	filter           PacketFilter
```
Comment (asticode) [inline packet_buffer.go:14]: Could you rename this variable from `filter` to `s` ?

File: packet_buffer.go [hunk 7]
```diff
@@ -11,15 +11,17 @@ import (
...
 	r                io.Reader
 	packetReadBuffer []byte
 }
 
 // newPacketBuffer creates a new packet buffer
-func newPacketBuffer(r io.Reader, packetSize int) (pb *packetBuffer, err error) {
+func newPacketBuffer(r io.Reader, packetSize int, filter PacketFilter) (pb *packetBuffer, err error) {
```
Comment (asticode) [inline packet_buffer.go:20]: Could you rename this variable from `filter` to `s` ?

File: packet_buffer.go [hunk 8]
```diff
@@ -121,19 +123,23 @@ func (pb *packetBuffer) next() (p *Packet, err error) {
...
 		}
-		return
-	}
 
-	// Parse packet
-	if p, err = parsePacket(astikit.NewBytesIterator(pb.packetReadBuffer)); err != nil {
-		err = fmt.Errorf("astits: building packet failed: %w", err)
-		return
+		// Parse packet
+		if p, err = parsePacket(astikit.NewBytesIterator(pb.packetReadBuffer), pb.filter); err != nil {
```
Comment (asticode) [inline packet_buffer.go:138]: You'll need to add `&& !errors.Is(errSkippedPacket)` after `err != nil`

File: packet_test.go [hunk 9]
```diff
@@ -45,12 +45,12 @@ func TestParsePacket(t *testing.T) {
...
+	_, err := parsePacket(astikit.NewBytesIterator(buf.Bytes()), nil)
 	assert.EqualError(t, err, ErrPacketMustStartWithASyncByte.Error())
 
 	// Valid
 	b, ep := packet(packetHeader, *packetAdaptationField, []byte("payload"), true)
-	p, err := parsePacket(astikit.NewBytesIterator(b))
+	p, err := parsePacket(astikit.NewBytesIterator(b), nil)
 	assert.NoError(t, err)
 	assert.Equal(t, p, ep)
 }
```
Comment (asticode) [inline packet_test.go:56]: Could you add a test for the filter here?

Something like:

```go
_, err = parsePacket(astikit.NewBytesIterator(b), func(p *Packet) bool { return true })
assert.EqualError(t, err, ...)
```

**Review by asticode — CHANGES_REQUESTED** [review]

File: packet.go (hunk 5 above)
Comment (k-danil) [reply packet.go:114]: Should we pass this error to NextData() caller or can we  just `continue` loop if such error occur?
In my opinion NextData() caller don't care much about transport part of protocol as long as data collected.

Comment (k-danil) [reply packet.go:114]: Or this error only for internal purpose between packetBuffer.next() and parsePacket()?

File: demuxer.go (hunk 1 above)
Comment (k-danil) [reply demuxer.go]: Done

File: demuxer.go (hunk 2 above)
Comment (k-danil) [reply demuxer.go]: Done

File: demuxer.go (hunk 3 above)
Comment (k-danil) [reply demuxer.go]: Done

File: packet.go (hunk 4 above)
Comment (k-danil) [reply packet.go]: Done

File: packet_buffer.go (hunk 6 above)
Comment (k-danil) [reply packet_buffer.go]: Done

File: packet_buffer.go (hunk 7 above)
Comment (k-danil) [reply packet_buffer.go]: Done

File: packet_buffer.go (hunk 8 above)
Comment (k-danil) [reply packet_buffer.go]: Fixed with additional `if` for the sake of line shortening

File: packet_test.go (hunk 9 above)
Comment (k-danil) [reply packet_test.go:60]: Done. Also added test case for TestDemuxerNew.

File: demuxer.go [hunk 10]
```diff
@@ -23,21 +23,27 @@ var (
...
+	packetPool   *packetPool
+	programMap   *programMap
+	r            io.Reader
 }
 
 // PacketsParser represents an object capable of parsing a set of packets containing a unique payload spanning over those packets
 // Use the skip returned argument to indicate whether the default process should still be executed on the set of packets
 type PacketsParser func(ps []*Packet) (ds []*DemuxerData, skip bool, err error)
 
+// Use this option if you need to filter out unwanted packets from your pipeline. NextPacket() will return the next unfiltered packet if any.
```
Comment (asticode) [inline demuxer.go:44]: Mmm I messed up in my previous comment and the comment I asked you to update this with was not fully written. Could you replace the comment with this:

```
// PacketSkipper represents an object capable of skipping a packet before parsing its payload. Its header and adaptation field is parsed and provided to the object.
// Use this option if you need to filter out unwanted packets from your pipeline. NextPacket() will return the next unskipped packet if any.
```

File: demuxer.go [hunk 11]
```diff
@@ -78,6 +84,13 @@ func DemuxerOptPacketsParser(p PacketsParser) func(*Demuxer) {
 	}
 }
 
+// DemuxerOptPacketSkipper returns the option to set the packet filter
```
Comment (asticode) [inline demuxer.go:87]: Could you update this comment to `// DemuxerOptPacketSkipper returns the option to set the packet skipper`

File: packet.go [hunk 12]
```diff
@@ -109,6 +112,11 @@ func parsePacket(i *astikit.BytesIterator) (p *Packet, err error) {
 		}
 	}
 
+	// Use optional packet filter
```
Comment (asticode) [inline packet.go:115]: Could you replace this comment with `// Skip packet`?

File: packet_buffer.go [hunk 13]
```diff
@@ -121,19 +124,25 @@ func (pb *packetBuffer) next() (p *Packet, err error) {
 		pb.packetReadBuffer = make([]byte, pb.packetSize)
 	}
 
-	if _, err = io.ReadFull(pb.r, pb.packetReadBuffer); err != nil {
-		if err == io.EOF || err == io.ErrUnexpectedEOF {
-			err = ErrNoMorePackets
-		} else {
-			err = fmt.Errorf("astits: reading %d bytes failed: %w", pb.packetSize, err)
+	// We must return some results to user even if current packet has been filtered
```
Comment (asticode) [inline packet_buffer.go:127]: Could you replace this comment with `// Loop to make sure we return a packet even if first packets are skipped`?

**Review by asticode — CHANGES_REQUESTED** [review]
Thanks for the previous changes, only minor comments changes needed now 👍

File: packet.go (hunk 5 above)
Comment (asticode) [reply packet.go]: You can keep it as it is, in my mind this error was only intended to the caller of `parsePacket()`

File: packet_test.go (hunk 9 above)
Comment (asticode) [reply packet_test.go:60]: 👍

File: demuxer.go (hunk 10 above)
Comment (k-danil) [reply demuxer.go]: Done

File: demuxer.go (hunk 11 above)
Comment (k-danil) [reply demuxer.go]: Done

File: packet.go (hunk 12 above)
Comment (k-danil) [reply packet.go]: Done

File: packet_buffer.go (hunk 13 above)
Comment (k-danil) [reply packet_buffer.go]: Done

**Comment by asticode** [comment]
No worries, cheers! ❤️

**Pull request merged**